Feature/product

ehghksvjscl opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 로그인 버튼 활성화 기능 구현(예시)
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- id '@' 포함, pw 5 글자 이상 입력 시 버튼 활성화(예시)

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
- 함수 로직이 너무 긴 것 같습니다. 좀 더 효율적인 방법이 없을지 궁금합니다.(예시) 

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

ehghksvjscl committed:
Create product app

ehghksvjscl committed:
New
MainView add
AllView add
Product modeling

ehghksvjscl committed:
1

ehghksvjscl committed:
2

ehghksvjscl committed:
New
Product_Star models
Product_Plag models Add

ehghksvjscl committed:
MOD
Product 모델 작성
MainView,AllView,DetailView 작성(매인 페이지,모든 페이지,상세페이지)
UnitTest 진행중

evergreen-david commented on README.md:
```diff
@@ -5,5 +5,10 @@
+- [] UnitTest 진행중
 
+
+URL : https://aquerytool.com:443/aquerymain/index/?rurl=5ec036c5-24bd-44c0-aea4-ab376e04c6c5&
+Password : 1pa0p1
```
URL, Password 를 이렇게 노출하면 안됩니다. 삭제 부탁 드리며 정리 부탁 드립니다.

evergreen-david commented on products/models.py:
```diff
@@ -0,0 +1,142 @@
+    def __str__(self):
+        return self.name
+
+class Product_Star(models.Model):
+    star_5 = models.IntegerField(default=0)
+    star_4 = models.IntegerField(default=0)
+    star_3 = models.IntegerField(default=0)
+    star_2 = models.IntegerField(default=0)
+    star_1 = models.IntegerField(default=0)
```
점수를 카운트하는 컬럼으로 보이는데요. 이렇게 개수를 별도로 저장하는것이 아니고 상품에 대하여 필터링 후에 평점를 카운트하여 사용해야 합니다.

ehghksvjscl replied:
해당 필드는 점수 카운터 용이 아니고 조회용입니다.

ehghksvjscl replied:
Product는 조회용 필드입니다.

인기순으로 조회 라는 기준을 맞추기 위해 각 상품당 별점들의 개수를 크롤링 하는 작업을 하였습니다.
평점으로 크롤링해서 데이터를 넣을시 평점 5.0 상품이 너무 많은 관계로 차별을 두고자 하였습니다.
리뷰 크롤링은 하나의 제품당 30개 정도를 크롤링 했으며, 각 리뷰의 평점이 1~5까지 int로 되어 있고 5점인 리뷰가 상당히 많기 떄문에 변별력이 없어서 이와 같이 하였습니다.

이와 같은 상황에는 어떻게 처리를 하는게 좋을까요?

evergreen-david commented on products/urls.py:
```diff
@@ -0,0 +1,9 @@
+
+urlpatterns = [
+    path("main", views.MainView.as_view()),
+    path("all", views.AllView.as_view()),
```
product로 수정 부탁 드립니다. 그리고 모든 상품을 리턴하게 되면 상품이 수천, 수만개 이상이면 리턴하는데 성능상 문제가 발생하기 때문에 쿼리스트링으로 받는 상품의 개수를 받아서 10개또는 20개정도만 리턴하도록 수정을 부탁 드립니다.

ehghksvjscl replied:
product로 수정 부탁 드립니다.  (해당 요구사항이 아래 처럼 바꾸라는 말씀이신가요?)
path("main", views.MainView.as_view()) -> path("main", views.ProductMainView.as_view())
 path("all", views.AllView.as_view()), ->  path("all", views.ProductAllView.as_view()),

evergreen-david commented on products/tests.py:
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
```
엔드포인트랑 1개씩 대표적인 success 케이스만 유닛 테스트 수행 부탁 드립니다.
즉 아래 케이스면 main, all, detail 에 대해서 1개씩 유닛 테스트가 있어야 합니다.

ehghksvjscl committed:
d

ehghksvjscl committed:
유닛테스트 작성

ehghksvjscl committed:
Product View이름 변경

ehghksvjscl committed:
MOD
prodcut_star 삭제 -> product에 star_average 추가(실수형)

soheon-lee commented on products/models.py:
```diff
@@ -0,0 +1,137 @@
+
+
+class Product_Plag(models.Model):
+    flag_sale = models.CharField(max_length=10)
+    flag_gift = models.CharField(max_length=10)
+    flag_new = models.CharField(max_length=10)
+    flag_best = models.CharField(max_length=10)
```
`=` 기준으로 간격 맞춰주세요

soheon-lee commented on products/models.py:
```diff
@@ -0,0 +1,137 @@
+        return self.name
+
+
+class Product_Plag(models.Model):
```
Flag가 아니라 Plag 인가요?

soheon-lee commented on products/models.py:
```diff
@@ -0,0 +1,137 @@
+        db_table = "image"
+
+
+class Product_Detail_Image(models.Model):
```
클래스 이름에는 언더바를 사용하지 않습니다. 대문자로 띄어쓰기가 구분되니 언더바는 삭제해주세요
```suggestion
class ProductDetailImage(models.Model):
```

soheon-lee commented on products/models.py:
```diff
@@ -0,0 +1,137 @@
+
+
+class Product(core_models.TempDate):
+    skin = models.ForeignKey("Skin_Category", on_delete=models.CASCADE, null=True)
+    genre = models.ForeignKey("Genre_Category", on_delete=models.CASCADE, null=True)
+    line = models.ForeignKey("Line_Category", on_delete=models.CASCADE, null=True)
+    online = models.ForeignKey("Online_Category", on_delete=models.CASCADE, null=True)
+    flag = models.ForeignKey("Product_Plag", on_delete=models.CASCADE, null=True)
+    product_detail = models.ForeignKey("Product_Detail", on_delete=models.CASCADE, null=True)
+    star_average = models.DecimalField(max_digits=5, decimal_places=2, default=0)
+    name = models.CharField(max_length=100)
+    main_image_url = models.CharField(max_length=1000, default="")
+    is_main = models.BooleanField(default=False)
```
`=` 기준으로 정렬해주세요

soheon-lee commented on products/models.py:
```diff
@@ -0,0 +1,137 @@
+
+
+class Distinct_Product(core_models.TempDate):
+    skin = models.ForeignKey("Skin_Category", on_delete=models.CASCADE, null=True)
+    genre = models.ForeignKey("Genre_Category", on_delete=models.CASCADE, null=True)
+    line = models.ForeignKey("Line_Category", on_delete=models.CASCADE, null=True)
+    online = models.ForeignKey("Online_Category", on_delete=models.CASCADE, null=True)
+    flag = models.ForeignKey("Product_Plag", on_delete=models.CASCADE, null=True)
+    name = models.CharField(max_length=100)
```
`=` 기준으로 정렬해주세요

soheon-lee commented on products/models.py:
```diff
@@ -0,0 +1,137 @@
+
+
+class Product_Detail(core_models.TempDate):
+    tag = models.CharField(max_length=100)
+    price = models.IntegerField()
+    price_sale = models.IntegerField()
```
`=` 기준으로 정렬해주세요

soheon-lee commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+from django.http import JsonResponse
+
+from . import models
+from reviews import models as reviews_models
```
```suggestion
from reviews.models import Review
```
사용할 클래스를 바로 import 해주세요

soheon-lee commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+from django.views import View
+from django.http import JsonResponse
+
+from . import models
```
```suggestion
from .models import (
    Product,
    ProductDetailImage
)
```

models에서 사용할 클래스를 하나씩 import 하시는 것을 추천드립니다.

soheon-lee commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+
+class MainView(View):
+    def get(self, request):
+        data = models.Product.objects.filter(is_main=True).values(
```
`Product`클래스를 바로 import 하셨다면, 클래스를 바로 사용하실 수 있습니다.
```suggestion
        data = Product.objects.filter(is_main=True).values(
```

soheon-lee commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+            "star_average",
+        )
+        reviews = (
+            reviews_models.Review.objects.exclude(image_url="")
```
```suggestion
            Review.objects.exclude(image_url="")
```

soheon-lee commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+        return JsonResponse({"data": list(data), "reviews": list(reviews)}, status=200)
+
+
+class AllView(View):
```
클래스 이름이 너무 모호합니다. 모든 제품의 리스를 보여주는 뷰이므로 그 의미가 보다 잘 전달되도록 `ProductListView`등을 추천드립니다.

soheon-lee commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+
+class AllView(View):
+    def get(self, request):
+        data = models.Product.objects.all().values(
```
`all( )` 메소드와 `values( )` 메소드는 같이 사용하지 않으셔도 됩니다 필요한 values만 쓰도록 해주세요.

soheon-lee commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+        return JsonResponse({"data": list(data)}, status=200)
+
+
+class DetailView(View):
```
역할을 보다 더 잘 알 수 있도록 `ProductDetailVeiw`등으로 수정 부탁드립니다.

soheon-lee commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+class DetailView(View):
+    def get(self, request, pk):
+        review_list = []
+        detail_pk = models.Product.objects.filter(id=pk)
+        image_pk = models.Product_Detail_Image.objects.filter(product_detail_id=pk)
+        images = image_pk.values("image__image_url")
+        data = detail_pk.values("product_detail__detail_html")
+        reviews = (
```
`=`을 기준으로 정렬해주세요

soheon-lee commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+        images = image_pk.values("image__image_url")
+        data = detail_pk.values("product_detail__detail_html")
+        reviews = (
+            reviews_models.Review.objects.select_related("worry", "skintype")
```
```suggestion
            Review.objects.select_related("worry", "skintype")
```

soheon-lee commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+class DetailView(View):
+    def get(self, request, pk):
+        review_list = []
+        detail_pk = models.Product.objects.filter(id=pk)
+        image_pk = models.Product_Detail_Image.objects.filter(product_detail_id=pk)
```
```suggestion
        detail_pk = Product.objects.filter(id=pk)
        image_pk = Product_Detail_Image.objects.filter(product_detail_id=pk)
```

soheon-lee commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+
+class DetailView(View):
+    def get(self, request, pk):
+        review_list = []
```
사용하지 않는 리스트를 선언하셨습니다. 제거해주세요.

soheon-lee commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+
+
+class DetailView(View):
+    def get(self, request, pk):
```
변수의 이름은 그 역할과 담는 데이터를 직관적으로 알 수 있도록 설정해주세요. `pk`보다는 `product_id`를 추천드립니다.

soheon-lee commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+            )
+        )
+        return JsonResponse(
+            {"images": list(images), "datas": list(data), "reviews": list(reviews)},
```
```suggestion
            {"images": list(images), "data": list(data), "reviews": list(reviews)},
```

ehghksvjscl committed:
Mod
models.py 줄간격 정렬 완료

ehghksvjscl committed:
수정
Product_Plag -> Product_Flag 이름 변경

ehghksvjscl committed:
수정
Class이름에 _제거

ehghksvjscl committed:
수정
models에서 사용할 클래스 하나씩 import 하기

ehghksvjscl committed:
수정
모호한 클레스 이름 변경
MainView -> ProductMainView
AllView -> ProductAllView
Detail -> ProductDetailView

ehghksvjscl committed:
datas -> data 딕셔너리 키 이름 변경 후 프론트 전달

ehghksvjscl committed:
Mod
all( ) 메소드와 values( ) 메소드는 같이 사용하지 않습니다. 수정 완료

ehghksvjscl committed:
Mod
AllView -> ProductListView 클레스 이름 변경

ehghksvjscl committed:
사용 안하는 변수 제거

evergreen-david commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+
+class ProductDetailView(View):
+    def get(self, request, pk):
+        detail_pk = Product.objects.filter(id=pk)
+        image_pk = ProductDetailImage.objects.filter(product_detail_id=pk)
```
Product 와 ProductDetailImage 간에 관계가있는데 객체간 관계를 이용해서 쿼리 해보시면 좋을 것 같습니다.

evergreen-david commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+        reviews = (
+            reviews_Review.objects.select_related("worry", "skintype")
+            .filter(product_id=pk)
+            .values(
+                "star_point", "worry__name", "skintype__name", "content", "image_url"
```
values 로 가져오지말고 위에서 같이 select_related 를 사용해서 쿼리해서 리스트 컴프리헨션으로 리턴하면 가독성도 훨씬 좋아집니다.

evergreen-david commented on products/views.py:
```diff
@@ -0,0 +1,72 @@
+
+class ProductListView(View):
+    def get(self, request):
+        data = Product.objects.prefetch_related("flag","product_detail").values(
```
여기도 values 로 하지말고 쿼리한 결과를 리스트로 리턴하기 때문에 리스트 컴프리헨션으로 바꾸면 좋을것같습니다.

evergreen-david commented on products/tests.py:
```diff
@@ -0,0 +1,34 @@
+
+
+class ProductAllTest(TestCase):
+    def test_product_all_get_success(self):
```
setup, teardown 이 없는데 유닛테스트 확인 부탁 드립니다.

ehghksvjscl committed:
MOD
UnitTest 수정 완료
